src/util/memstream.rs

```diff
@@ -0,0 +1,163 @@
+// Copyright 2018 MaidSafe.net limited.
+//
+// This SAFE Network Software is licensed to you under the MIT license <LICENSE-MIT
+// https://opensource.org/licenses/MIT> or the Modified BSD license <LICENSE-BSD
+// https://opensource.org/licenses/BSD-3-Clause>, at your option. This file may not be copied,
+// modified, or distributed except according to those terms. Please review the Licences for the
+// specific language governing permissions and limitations relating to use of the SAFE Network
+// Software.
+
+//! Tokio compatible in-memory stream implementation. Helpful for testing.
+
+// TODO(povilas): move to future_utils crate - this module is generic and highly reusable.
+
+use priv_prelude::*;
+use std::cmp;
+use std::io::{self, Read, Write};
+
+/// In-memory stream implementing `AsyncWrite + AsyncRead`. This stream is not connected with
+/// any other stream. Meaning what you write to the stream you can later read from it.
+#[derive(Default, Debug)]
+pub struct EchoStream {
+    buf: Vec<u8>,
+    shutdown: bool,
+}
+
+impl Read for EchoStream {
+    /// Reads as much data as possible to the given buffer and removes that data from internal
+    /// buffer.
+    fn read(&mut self, buf: &mut [u8]) -> io::Result<usize> {
+        if self.buf.is_empty() {
+            if self.shutdown {
+                return Ok(0);
+            } else {
+                return Err(io::ErrorKind::WouldBlock.into());
+            }
+        }
+
+        let bytes_to_read = cmp::min(self.buf.len(), buf.len());
+        buf.iter_mut()
+            .zip(self.buf.drain(..bytes_to_read))
+            .for_each(|(dst, src)| *dst = src);
+        Ok(bytes_to_read)
+    }
+}
+
+impl Write for EchoStream {
+    fn write(&mut self, buf: &[u8]) -> io::Result<usize> {
+        self.buf.extend_from_slice(buf);
+        Ok(buf.len())
+    }
+
+    fn flush(&mut self) -> io::Result<()> {
+        Ok(())
+    }
+}
+
+impl AsyncRead for EchoStream {}
+
+impl AsyncWrite for EchoStream {
+    /// Analogous to sending FIN to myself.
+    /// This will unblock `read()` calls which will get 0 as result value.
+    fn shutdown(&mut self) -> io::Result<Async<()>> {
+        self.shutdown = true;
+        Ok(Async::Ready(()))
+    }
+}
+
+#[cfg(test)]
+mod tests {
+    use super::*;
+
+    mod memstream {
```

Comment:
Is this necessary ? The tests are already under the module called `memstream` (i.e., this file)

Reply:
oops, meant `mod echostream`. This nests test cases related only to this structure, so if there were other structures or global functions, they wouldn't go into this module. I'll rebase the #975.